exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/retry/RetryInterceptor.java

```diff
@@ -38,44 +47,72 @@ public RetryInterceptor(RetryPolicy retryPolicy, Function<Response, Boolean> isR
-    if (!Boolean.TRUE.equals(isRetryable.apply(response))) {
-      return response;
-    }
+    Response response = null;
+    IOException exception = null;
+    int attempt = 0;
     long nextBackoffNanos = retryPolicy.getInitialBackoff().toNanos();
-    // Made an attempt already, starting with the 2nd.
-    for (int attempt = 2; attempt <= retryPolicy.getMaxAttempts(); attempt++) {
-      // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exponential-backoff
-      // Careful - the document says "retry attempt" but our maxAttempts includes the original
-      // request. The numbers are tricky but the unit test should be more clear.
-      long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
-      long backoffNanos = randomLong.get(upperBoundNanos);
+    do {
+      if (attempt > 0) {
+        // Compute and sleep for backoff
+        // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exponential-backoff
+        long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
+        long backoffNanos = randomLong.get(upperBoundNanos);
+        nextBackoffNanos = (long) (nextBackoffNanos * retryPolicy.getBackoffMultiplier());
+        try {
+          sleeper.sleep(backoffNanos);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          break; // Break out and return response or throw
+        }
+        // Close response from previous attempt
+        if (response != null) {
+          response.close();
+        }
+      }
 
+      attempt++;
       try {
-        sleeper.sleep(backoffNanos);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        return response;
+        response = chain.proceed(chain.request());
+      } catch (IOException e) {
+        throttlingLogger.log(
+            Level.FINE,
+            "Error executing "
+                + chain.request().method()
+                + " "
+                + chain.request().url()
+                + ": "
+                + e.getMessage());
+        exception = e;
       }
-      // Close response from previous attempt and try again.
-      response.close();
-      response = chain.proceed(chain.request());
-      if (!Boolean.TRUE.equals(isRetryable.apply(response))) {
+      if (response != null && !Boolean.TRUE.equals(isRetryable.apply(response))) {
         return response;
       }
-      nextBackoffNanos = (long) (nextBackoffNanos * retryPolicy.getBackoffMultiplier());
+      if (exception != null && !Boolean.TRUE.equals(isRetryableException.apply(exception))) {
+        throw exception;
+      }
+    } while (attempt < retryPolicy.getMaxAttempts());
+
+    if (response != null) {
+      return response;
     }
-    return response;
+    throw exception;
+  }
+
+  // Visible for testing
+  static boolean isRetryableException(IOException e) {
+    return e instanceof SocketTimeoutException;
```

Comment:
I think we're not supposed to retry on readTimeout because all the data has already been sent and may have been processed by the server. Well it's my interpretation of the spirit of the spec. Do you know if we can determine what type of timeout it is from some detail on the exception? `instanceof` seems like it will hold for connect, read, and write, though not call

Reply:
My read of the spec is that the read / write timeout behavior is unspecified, but that retrying is expected and acceptable. The relevant bit is [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#duplicate-data). When we write a request and a read timeout occurs waiting for a response, we don't know whether the server successfully processed the bytes or not. Better to guarantee delivery as the risk of duplicate data. 

If we wanted to have behavior that differs for read / write / connect timeout, its possible to do so by inspecting the `SocketTimeoutException` error message:
- For connect timeouts, error message is "Connect timed out"
- For read timeouts, error message is "Read timed out"
- For write timeouts, error message is "timeout"

Not super confident that the messages are reliable though because I've seen at least the case of the message vary by JVM. I.e. in one it was "Connect timed out", in another it was "connect timed out".

Reply:
I guess it's OK - the reason I still have some concern as readTimeout specifically may not be a transient error. If the server always takes too long to return its response for the telemetry we send, we will always retry until max attempts all the time. This generally indicates either the timeout is too low or a bigger problem in general, so this is probably still OK but something to keep in mind. It wouldn't be OK if it caused a queue to grow unbounded somewhere, but I don't think we have that behavior with BSP, would be good if you can doublecheck.

Reply:
> It wouldn't be OK if it caused a queue to grow unbounded somewhere, but I don't think we have that behavior with BSP, would be good if you can doublecheck that retries will not amplify a bad effect on the app itself.

Yeah the BSP will block for a configurable amount of time waiting for the export to complete, and the export itself has a configurable timeout that it can spend waiting for the export, which consists at least one attempt and possibly more if retry is enabled and the timeout allows it. If the exporter timeout is longer than the BSP timeout, then an export may take longer than the BSP is willing to wait, and the BSP may start another export before the first is finished. In this scenario, its not clear to me what happens when one of the grpc transports (i.e. netty, netty-shadd, or okhttp) is used. But when the okhttp client is used, we use `OkHttpUtil.newDispatcher`, which uses a thread pool with a max size of `Integer.MAX_VALUE`. 

So if many exports end up in flight simultaneously, there could be congestion that will slow things down but its not like the exports are backing up in a queue. However, this does underscore that its is important to pay attention to the settings for `OTEL_BSP_EXPORT_TIMEOUT` and `OTEL_EXPORTER_OTLP_TIMEOUT` together. I wouldn't advise having a `OTEL_EXPORTER_OTLP_TIMEOUT` that is greater than `OTEL_BSP_EXPORT_TIMEOUT`. 

But I don't feel strongly about the read timeout stuff. I would be ok checking that the exception has a message contains "Connect timed out" as criteria to retry. Can always make the check more broad in the future.